src/system.jl

```diff
@@ -0,0 +1,302 @@
+"""
+    ServicesTimeSeries(reg, spin, sup_on, sup_off)
+
+Type defining services time series.  Fields are KeyedArray where the keys are generator names
+x datetimes.  Only generators that provide each service are included in the array.
+"""
+struct ServicesTimeSeries
+    reg::KeyedArray{Float64}
+    spin::KeyedArray{Float64}
+    sup_on::KeyedArray{Float64}
+    sup_off::KeyedArray{Float64}
+end
+
+"""
+    Zone(number, reg, spin, sup_on, sup_off)
+
+Type defining a market zone.  The `Zone` is identified by a number.  The other fields contain
+the service requirements for the zone.
+"""
+struct Zone
+    number::Int64
+    reg::Float64
+    spin::Float64
+    sup_on::Float64
+    sup_off::Float64
+end
+
+###### Static Component Types ######
+
+abstract type StaticComponent end
+
+Base.length(components::StaticComponent) = length(getfield(components, 1))
+# define interfaces? Iterator? Table?
+
+"""
+    Generators(
+        name::Vector{Int}
+        zone::Vector{Int}
+        startup_cost::Vector{Float64}
+        shutdown_cost::Vector{Float64}
+        no_load_cost::Vector{Float64}
+        time_at_status::Vector{Float64}
+        min_uptime::Vector{Float64}
+        min_downtime::Vector{Float64}
+        ramp_up::Vector{Float64}
+        ramp_down::Vector{Float64}
+        initial_gen::Vector{Float64}
+        technology::Vector{Symbol}
+    )
+
+Type for static generator component attributes (i.e. things that describe a generator that
+are not time series data).
+"""
+struct Generators <: StaticComponent
+    name::Vector{Int}
+    zone::Vector{Int}
+    startup_cost::Vector{Float64}
+    shutdown_cost::Vector{Float64}
+    no_load_cost::Vector{Float64}
+    time_at_status::Vector{Float64}
```

Comment:
`time_at_status` should probably be a time-series

Reply:
>`time_at_status` should probably be a time-series

Can you explain why would that be? `time_at_status` indicates for how long the unit is in its current state (on/off) at the first time period. We don't know the time at status for all time periods since that will be defined as part of the optimisation (we know this if we have the commitment statuses, e.g. if we're running economic dispatch, but then `time_at_status` isn't used at all and it's also trivial to compute)

Reply:
because it is not static; it changes each day. it will have 1 daily value (for the first hour of the day), rather than 24 hourly values, but it's still a changing value, unlike everything else in this struct

Reply:
Sorry, it might be because I'm a bit out of the loop – does it matter if something changes across days? Isn't one of the basic assumptions that a system is defined for a given day? Or are we aiming to be as general as possible here?

Also, it's true that start-up/shutdown/no-load costs _should not_ change on a daily basis, but they _could_ change sometimes (not under our current framework, but potentially in the future)

Reply:
> does it matter if something changes across days? Isn't one of the basic assumptions that a system is defined for a given day? 

yeah, i think either option is fine, since we will a `System` will model a single day

But having things that are in practice unchanged from day-to-day opens up the possibility of reusing the same StaticComponent across Systems (if that's something we might ever want to do)

Reply:
I've put `time_at_status` and `initial_generation` in with the time series data.  I think it makes sense to treat them as time series although they will both only have one datetime dimension.  

The other thing that links them is they represent information we pass from one model to the next with [the logic in FNSims](https://gitlab.invenia.ca/invenia/research/FullNetworkSimulations.jl/-/blob/1bd229c7db0e9bb702b59ff1d5f8bdfc03036e14/src/simulation.jl#L508).  The time series arrays are a convenient format to perform those `_initial_status` and `_update_system_generation` operations on.